Delegate and block producer clients.

romfrolov opened the pull request:
Summary:
- delegate client
- block producer client

romfrolov committed:
Summary:
- Added skip for zero addresses when initating state from genesis block.
- TODO list for block producer client.
- TODO list for delegate client.

romfrolov committed:
Initial logic for block producer and delegate.

romfrolov committed:
Decoupled block production function.

romfrolov committed:
Refactored account module.

romfrolov committed:
Added to do notes and questions.

romfrolov committed:
Added block validation.

romfrolov committed:
Merge remote-tracking branch 'origin' into delegates

romfrolov committed:
Fixed invalid type of locked amount. Supressed log in test.

romfrolov committed:
Moved constants to lib.

romfrolov committed:
Fixed file case (probably).

romfrolov committed:
Added validation of block producer.

romfrolov committed:
Completed some clauses of TODO notes.

romfrolov committed:
Added note on "Random number distribution".

romfrolov committed:
Extended block producer prototype (certificate selection) logic.

damirka commented on core/account.js:
```diff
@@ -0,0 +1,138 @@
+
+    while (!secretKey || !secp256k1.privateKeyVerify(secretKey)) {
+        secretKey = randomBytes(32);
+    }
```
Генерацию secret key я бы перенёс в отдельную функцию типа "generatePrivateKey". Кажется логичным.

damirka commented on core/blockchain.js:
```diff
@@ -0,0 +1,189 @@
+ * @return {Object}
+ */
+function handleStake(tx, block) {
+    const sender = block.state.find(account => account.address === tx.from);
```
Best practice - оборачивать boolean выражения в круглые скобки i.e. ``` => (account.address === tx.from)```
Отображает операцию и позволяет читать стрелочные функции как "действие => результат".

damirka commented on core/delegate.js:
```diff
@@ -0,0 +1,95 @@
+        randomNumbers = [];
+    }
+
+    randomNumber = Math.random();
```
It may be better if didn't use decimals in random. Math.random() returns value in range 0-1.

damirka replied:
So it's something like (Math.random() * 10).toFixed(0)

romfrolov replied:
Sure. Random generation was just mocked out.

romfrolov committed:
Moved secret key generation to helpers lib.

romfrolov committed:
Sync with master branch.

romfrolov committed:
Added account module back.

damirka commented on core/block-producer.js:
```diff
@@ -0,0 +1,67 @@
+ *
+ * @type {String}
+ */
+const DELEGATES_GROUP = 'delegates';
```
= exports.DELEGATES_GROUP =

damirka commented on core/blockchain.js:
```diff
@@ -0,0 +1,189 @@
+ * @return {Object}
+ */
+exports.handleTransaction = function handleTransaction(tx, block) {
+    if (isVote(tx.data)) {
```
switch (true) is so true.

romfrolov replied:
It just spawns a lot of `breaks`. Don't think it is worthy in this particular case.

damirka replied:
Instead of break there gonna be return statement which is even more clear.

damirka commented on core/blockchain.js:
```diff
@@ -0,0 +1,189 @@
+    let i = 0;
+
+    for (const account of block.state) {
+        for (let _ = 0; _ < account.certificates.length; _++) {
```
Oi, gryazno. Nikakogo go-style, while in javascript.

damirka replied:
It's i or any other letter. Underscore in JS means utility library.

damirka commented on core/blockchain.js:
```diff
@@ -0,0 +1,189 @@
+ * @param  {Object}  block
+ * @return {Object}
+ */
+exports.genesisAllocation = function genesisAllocation(genesisBlock, block) {
```
Name functions with verbs. 
```
function allocateGenesis() {}
// ... or ...
function getGenesisBlock() {}
```

damirka commented on core/delegate.js:
```diff
@@ -0,0 +1,95 @@
+/**
+ * Delegate client.
+ *
+ * @module core/delegate
```
Not a core. Move somewhere else. It's not a component - it's a nearly-complete client app.

damirka commented on core/delegate.js:
```diff
@@ -0,0 +1,95 @@
+ */
+const DELEGATES_GROUP = 'delegates';
+
+const secretKey = '';
```
This is weird. Why?

romfrolov committed:
Added event listener to block producer. Added a note with a question regarding production of first blocks after genesis. Minor fixes during review.

damirka committed:
Moved delegate and bp to clients/, changed network/ to services/

- added sync logic in every of 2 clients
- moved sync to services/sync
- small changes there and here

romfrolov committed:
Merge pull request #13 from WingsDao/clients

Moved delegate and bp to clients/, changed network/ to services/

borispovod committed:
Sign and verify messages

- signing and verifying messages
- fixed empty lines, comments, changed functions logic
- documentation updates

romfrolov committed:
Added missing semicolons.

romfrolov committed:
Merge remote-tracking branch 'origin' into delegates

romfrolov committed:
Account votes was changed to array (set) of delegate addresses. Voting amount is the accounts balance at the particular block. If the balance decreases, so does a voting amount.'

romfrolov committed:
Added state verification to test.

damirka committed:
Clients (#21)

- improved bp.js - added vote and stake txs
- now bp continuously produces block with stake and vote txs
- improved core/file-peer to match new requirements
- fixed issue with non-closing when reached number of requests
- improved single string peering-pulling by removing Streams
- brand-new documentation
- updates on core/pool and chaindata modules
- chaindata now uses os.EOL for other operating systems
- pool.drain() now returns all the values from pool
- гpdated waiter to support Infinity waiting time
- now you can pass Infinity as timeout and Promise will finish only
by number of successful emissions for event
- runner setup
- clean up, docs updates, SECRET_KEY in dg and bp
- now process.env.SECRET_KEY is used in child processes
- however master-runner still requires updates
- changed way of attaching listeners in BP and DG
- added trial version of genesis library
- support writing genesis to file
- update in initiation of genesis state
- removed empty addresses from genesis.json
- fixed path in genesis
- update README.md
- added the contributors section to README
- updated file-peer to support server shutdown promise
- now 2 values are returned in file-peer: {port, promise}
- promise resolves with number of requests made and server timeout
- optimised listeners in clients/delegate
- better object assignment during genesis state initialization
- account generation and genesis initialisation in runner
- fixed incorrect type conversion of secret key in runner and delegate client
- delegate block verification 
- zero block check was added to block production to remove possibility of repeated allocation
- documentation for blockVerification method
- signed random number and message signature verification
- delegate client code improvement after review
- added signature and public key to streaming messsage
- fixed variable naming
- added wallet service to initialize account in one place
- also cleaned up delegate client
- documentation updates
- added block signature verification

romfrolov committed:
Removed resolved questions.

romfrolov committed:
Described two possible scenarios of selecting block producer.

romfrolov committed:
Fixed lexical error in block producer selection scenario.

romfrolov committed:
Producer selection algorithm. (#29)

- draft version of block producer selection algorithm
- supply final random number to block producer selection function
- added TOTAL_CERTIFICATES env setting for better prototyping

damirka committed:
Added block producer selection, increased number of BPs in runner

damirka committed:
Fixed locked (str + int) to (int + int)

damirka committed:
Added ping mechanics for address pinging (#32)

- added ping mechanics for address pinging

borispovod committed:
Core service (#30)

* Update README.md

Added the contributors section to README

* Method to get account address as hex implementation

* fixed initial state in genesis

* core service

* Fixed tabs, comments, spaces, changed 'getStrAddress' to 'getHexAddress',
getBlockProducers and isBlockProducer functions implementation

* account file

* merging chain and blockchain service to one, same with unit tests

* removed useless semicolon

* removed empty lines after function

* added state to genesis

* removed empty line

* updated tests for blockchain service

* Update to package.json to create folder 'data' after installation

* Empty line to blockchain.js, updated constants with active and successor delegates constants, removed usage of 'this' in blockchain.js

* empty line in the end of file

damirka committed:
Changed frn generation function, implemented cert number (#33)

damirka committed:
Added bank client and event to request coin (#31)

- Added bank client and event to request coin
- Used new getHexAddress method in bank

damirka committed:
Leveldb integration (#34)

- changed frn generation function, implemented cert number
- added LevelDB scratch for future development
- chaindata is replaced with core/db
- removed core/pool, replaced it with core/db
- now core/db is the only point of access to data
- few bugs in stream pulling were fixed and now allow async writing
- fewer console logs
- improved lib/genesis to support multi-role records
- now both addProducer and addDelegate can refer to same address
- removed lib/parser as it was only used in tests
- removed console logs in tests
- mkdir -p after installation, removed state from initial genesis, fixed genesis block state
- added adapter to generate state from genesis block

damirka reviewed (approved):
Yet it's not the best PR by code and architecture, it has to be merged for more comfrortable future development.
